modelling/src/main/java/org/axonframework/modelling/repository/InMemoryRepository.java

```diff
@@ -0,0 +1,156 @@
+     * {@code idType}.
+     * <p>
+     * The repository uses a {@link ConcurrentHashMap} as the underlying storage mechanism and delegates entity
+     * lifecycle management to a {@link SimpleRepository}.
+     *
+     * @param idType     The type of the identifier for entities in this repository.
+     * @param entityType The type of entity stored in this repository.
+     */
+    public InMemoryRepository(@Nonnull Class<ID> idType, @Nonnull Class<E> entityType) {
+        requireNonNull(idType, "The idType may not be null");
+        requireNonNull(entityType, "The entityType may not be null");
+
+        this.storage = new ConcurrentHashMap<>();
+        this.delegate = new SimpleRepository<>(
+                idType,
+                entityType,
+                (identifier, processingContext) -> CompletableFuture.completedFuture(storage.get(identifier)),
+                (identifier, entity, processingContext) -> {
+                    if (entity != null) { // it's sometimes null!
+                        storage.put(identifier, entity);
+                    }
+                    return FutureUtils.emptyCompletedFuture();
+                }
+        );
+    }
+
+    @Override
+    public ManagedEntity<ID, E> attach(@Nonnull ManagedEntity<ID, E> entity,
+                                       @Nonnull ProcessingContext processingContext) {
+        return delegate.attach(entity, processingContext);
+    }
+
+    @Nonnull
+    @Override
+    public Class<E> entityType() {
+        return delegate.entityType();
+    }
+
+    @Nonnull
+    @Override
+    public Class<ID> idType() {
+        return delegate.idType();
+    }
+
+    @Override
+    public CompletableFuture<ManagedEntity<ID, E>> load(@Nonnull ID identifier,
+                                                        @Nonnull ProcessingContext processingContext) {
+        return delegate.load(identifier, processingContext);
+    }
+
+    @Override
+    public CompletableFuture<ManagedEntity<ID, E>> loadOrCreate(@Nonnull ID identifier,
+                                                                @Nonnull ProcessingContext processingContext) {
+        return delegate.loadOrCreate(identifier, processingContext);
+    }
+
+    @Override
+    public ManagedEntity<ID, E> persist(@Nonnull ID identifier, @Nonnull E entity,
+                                        @Nonnull ProcessingContext processingContext) {
+        return delegate.persist(identifier, entity, processingContext);
+    }
+
+    @Override
+    public void describeTo(@Nonnull ComponentDescriptor descriptor) {
+        descriptor.describeProperty("storageType", "ConcurrentHashMap");
+        descriptor.describeProperty("storageSize", storage.size());
+        delegate.describeTo(descriptor);
+    }
+
+    /**
+     * Returns the number of entities currently stored in this repository.
+     *
+     * @return The number of entities in storage.
+     */
+    public int size() {
```

Comment:
What are these size/clear/contains methods for? For testing? Perhaps we can make these `@Internal`? I think the entire repository has merit as a public API, but I am not so sure if these methods are reasonable to keep exposed indefinitely to our users.